app/services/users/find_or_create_from_provider_data.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module Users
+  class FindOrCreateFromProviderData
+    def initialize(provider_data:, feature_flag_id:)
+      @provider_data = provider_data
+      @feature_flag_id = feature_flag_id
+    end
+
+    attr_reader :provider_data, :feature_flag_id
+
+    def call
+      user = User.find_by(provider: provider_data.provider, uid: provider_data.uid, archived_at: nil)
+
+      if user
+        check_and_archive_clashing_user(provider_data.info.email, user)
+        user.assign_attributes(email: provider_data.info.email)
+      else
+        Rails.logger.info("[GAI] User not found using UID, UID=#{provider_data.uid}, using email to find user")
+        check_if_supplied_uid_matches_archived_account # CPDNPQ-2647
+        user = User.find_or_initialize_by(email: provider_data.info.email, archived_at: nil)
+        user.assign_attributes(provider: provider_data.provider, uid: provider_data.uid)
+      end
+
+      assign_provider_data(user, provider_data)
+      user.assign_attributes(feature_flag_id: feature_flag_id)
+
+      user.tap(&:save)
```

Comment:
This is the only part I'm not sure about, I know we inherited it like it but feels likely to result in a silent fail

I wonder if this would be better as `user.tap(&:save!)`

Reply:
This service is only called in one place, in `OmniauthController`.
Right after calling this service, the controller does it's own save, and sends the error to sentry - this is useful to see when unexpected errors are occurring.
The user is then just redirected to the `failed_sign_in_path`.